File: multi_agent/projects/im-desktop/src/services/profile_service.ts
/**
 * 用户资料服务 - 桌面端
 * 负责用户资料、在线状态、好友备注的获取和更新
 */

import { WebSocketService } from './websocket';
import { ApiClient } from './api_client';

export interface UserProfile {
  userId: string;
  nickname: string;
  avatarUrl: string;
  signature: string;
  email?: string;
  phone?: string;
  gender?: string;
  birthday?: string;
  region?: string;
  language?: string;
  status: UserStatus;
  customStatus?: Record<string, unknown>;
  createdAt: string;
  updatedAt: string;
}

export type UserStatus = 'ONLINE' | 'AWAY' | 'BUSY' | 'DO_NOT_DISTURB' | 'INVISIBLE' | 'OFFLINE';

export interface FriendRemark {
  friendId: string;
  remark: string;
  groupName: string;
  tags: string[];
  updatedAt: string;
}

export interface FriendGroup {
  groupId: string;
  name: string;
  sortOrder: number;
  memberCount: number;
}

export interface ProfileUpdatePayload {
  nickname?: string;
  avatarUrl?: string;
  signature?: string;
  email?: string;
  phone?: string;
  gender?: string;
  birthday?: string;
  region?: string;
  language?: string;
  status?: UserStatus;
}

class ProfileService {
  private wsHandler: WebSocketService | null = null;
  private currentUserId: string | null = null;
  private onlineUsers: Set<string> = new Set();
  private statusListeners: Set<(data: StatusChangeEvent) => void> = new Set();

  setWebSocketHandler(ws: WebSocketService) {
    this.wsHandler = ws;
    this.wsHandler.on('status_change', (data: StatusChangeEvent) => {
      if (data.status === 'ONLINE') {
        this.onlineUsers.add(data.userId);
      } else {
        this.onlineUsers.delete(data.userId);
      }
      this.statusListeners.forEach(cb => cb(data));
    });
  }

  async getProfile(userId: string): Promise<UserProfile> {
    const resp = await ApiClient.get<UserProfile>(`/api/profile/${userId}`);
    return resp.data;
  }

  async getPublicProfile(userId: string): Promise<UserProfile> {
    const resp = await ApiClient.get<UserProfile>(`/api/profile/${userId}/public`);
    return resp.data;
  }

  async updateProfile(userId: string, payload: ProfileUpdatePayload): Promise<UserProfile> {
    const resp = await ApiClient.put<UserProfile>(`/api/profile/${userId}`, payload);
    return resp.data;
  }

  async updateAvatar(userId: string, avatarUrl: string): Promise<UserProfile> {
    const resp = await ApiClient.patch<UserProfile>(`/api/profile/${userId}/avatar`, { avatarUrl });
    return resp.data;
  }

  async updateNickname(userId: string, nickname: string): Promise<UserProfile> {
    const resp = await ApiClient.patch<UserProfile>(`/api/profile/${userId}/nickname`, { nickname });
    return resp.data;
  }

  async updateSignature(userId: string, signature: string): Promise<UserProfile> {
    const resp = await ApiClient.patch<UserProfile>(`/api/profile/${userId}/signature`, { signature });
    return resp.data;
  }

  async updateStatus(userId: string, status: UserStatus): Promise<UserProfile> {
    const resp = await ApiClient.patch<UserProfile>(`/api/profile/${userId}/status`, { status });
    this.wsHandler?.send({
      type: 'profile_update',
      action: 'status_update',
      status: status
    });
    return resp.data;
  }

  async searchUsers(keyword: string, limit = 20): Promise<UserProfile[]> {
    const resp = await ApiClient.post<UserProfile[]>('/api/profile/search', { keyword, limit });
    return resp.data;
  }

  async getProfiles(userIds: string[]): Promise<Record<string, UserProfile>> {
    const resp = await ApiClient.post<Record<string, UserProfile>>('/api/profile/batch', { userIds });
    return resp.data;
  }

  async getOnlineUsers(): Promise<string[]> {
    const resp = await ApiClient.get<string[]>('/api/profile/online');
    return resp.data;
  }

  isOnline(userId: string): boolean {
    return this.onlineUsers.has(userId);
  }

  onStatusChange(callback: (data: StatusChangeEvent) => void): () => void {
    this.statusListeners.add(callback);
    return () => this.statusListeners.delete(callback);
  }

  // 好友备注
  async setFriendRemark(userId: string, friendId: string, remark: string, groupName?: string, tags?: string[]): Promise<FriendRemark> {
    const resp = await ApiClient.put<FriendRemark>('/api/profile/friend-remark', {
      userId, friendId, remark, groupName: groupName || '', tags: tags || []
    });
    return resp.data;
  }

  async getFriendRemark(userId: string, friendId: string): Promise<FriendRemark | null> {
    try {
      const resp = await ApiClient.get<FriendRemark>(`/api/profile/${userId}/friend-remark/${friendId}`);
      return resp.data;
    } catch {
      return null;
    }
  }

  async getAllFriendRemarks(userId: string): Promise<FriendRemark[]> {
    const resp = await ApiClient.get<FriendRemark[]>(`/api/profile/${userId}/friend-remarks`);
    return resp.data;
  }

  async removeFriendRemark(userId: string, friendId: string): Promise<void> {
    await ApiClient.delete(`/api/profile/${userId}/friend-remark/${friendId}`);
  }

  // 好友分组
  async createFriendGroup(userId: string, name: string, sortOrder = 0): Promise<FriendGroup> {
    const resp = await ApiClient.post<FriendGroup>(`/api/profile/${userId}/friend-group`, { name, sortOrder });
    return resp.data;
  }

  async getFriendGroups(userId: string): Promise<FriendGroup[]> {
    const resp = await ApiClient.get<FriendGroup[]>(`/api/profile/${userId}/friend-groups`);
    return resp.data;
  }

  async deleteFriendGroup(userId: string, groupId: string): Promise<void> {
    await ApiClient.delete(`/api/profile/${userId}/friend-group/${groupId}`);
  }
}

export interface StatusChangeEvent {
  userId: string;
  status: UserStatus;
}

export const profileService = new ProfileService();

File: im-mobile/lib/services/profile_service.dart
/// 用户资料服务 - Flutter移动端

import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/profile.dart';

class ProfileService {
  static const String _baseUrl = 'http://localhost:8080/api';
  final String _token;

  ProfileService({String token = ''}) : _token = token;

  Map<String, String> get _headers => {
        'Content-Type': 'application/json',
        if (_token.isNotEmpty) 'Authorization': 'Bearer $_token',
      };

  Future<UserProfile> getProfile(String userId) async {
    final resp = await http.get(
      Uri.parse('$_baseUrl/profile/$userId'),
      headers: _headers,
    );
    return UserProfile.fromJson(jsonDecode(resp.body)['data']);
  }

  Future<UserProfile> updateProfile(String userId, Map<String, dynamic> payload) async {
    final resp = await http.put(
      Uri.parse('$_baseUrl/profile/$userId'),
      headers: _headers,
      body: jsonEncode(payload),
    );
    return UserProfile.fromJson(jsonDecode(resp.body)['data']);
  }

  Future<UserProfile> updateAvatar(String userId, String avatarUrl) async {
    final resp = await http.patch(
      Uri.parse('$_baseUrl/profile/$userId/avatar'),
      headers: _headers,
      body: jsonEncode({'avatarUrl': avatarUrl}),
    );
    return UserProfile.fromJson(jsonDecode(resp.body)['data']);
  }

  Future<UserProfile> updateNickname(String userId, String nickname) async {
    final resp = await http.patch(
      Uri.parse('$_baseUrl/profile/$userId/nickname'),
      headers: _headers,
      body: jsonEncode({'nickname': nickname}),
    );
    return UserProfile.fromJson(jsonDecode(resp.body)['data']);
  }

  Future<UserProfile> updateSignature(String userId, String signature) async {
    final resp = await http.patch(
      Uri.parse('$_baseUrl/profile/$userId/signature'),
      headers: _headers,
      body: jsonEncode({'signature': signature}),
    );
    return UserProfile.fromJson(jsonDecode(resp.body)['data']);
  }

  Future<UserProfile> updateStatus(String userId, UserStatus status) async {
    final resp = await http.patch(
      Uri.parse('$_baseUrl/profile/$userId/status'),
      headers: _headers,
      body: jsonEncode({'status': status.apiValue}),
    );
    return UserProfile.fromJson(jsonDecode(resp.body)['data']);
  }

  Future<List<UserProfile>> searchUsers(String keyword, {int limit = 20}) async {
    final resp = await http.post(
      Uri.parse('$_baseUrl/profile/search'),
      headers: _headers,
      body: jsonEncode({'keyword': keyword, 'limit': limit}),
    );
    final List<dynamic> data = jsonDecode(resp.body)['data'];
    return data.map((e) => UserProfile.fromJson(e)).toList();
  }

  Future<Map<String, UserProfile>> getProfiles(List<String> userIds) async {
    final resp = await http.post(
      Uri.parse('$_baseUrl/profile/batch'),
      headers: _headers,
      body: jsonEncode({'userIds': userIds}),
    );
    final Map<String, dynamic> data = jsonDecode(resp.body)['data'];
    return data.map((k, v) => MapEntry(k, UserProfile.fromJson(v)));
  }

  Future<List<String>> getOnlineUsers() async {
    final resp = await http.get(
      Uri.parse('$_baseUrl/profile/online'),
      headers: _headers,
    );
    return List<String>.from(jsonDecode(resp.body)['data']);
  }

  // 好友备注
  Future<FriendRemark> setFriendRemark(String userId, String friendId,
      {String remark = '', String groupName = '', List<String> tags = const []}) async {
    final resp = await http.put(
      Uri.parse('$_baseUrl/profile/friend-remark'),
      headers: _headers,
      body: jsonEncode({
        'userId': userId,
        'friendId': friendId,
        'remark': remark,
        'groupName': groupName,
        'tags': tags,
      }),
    );
    return FriendRemark.fromJson(jsonDecode(resp.body)['data']);
  }

  Future<FriendRemark?> getFriendRemark(String userId, String friendId) async {
    try {
      final resp = await http.get(
        Uri.parse('$_baseUrl/profile/$userId/friend-remark/$friendId'),
        headers: _headers,
      );
      return FriendRemark.fromJson(jsonDecode(resp.body)['data']);
    } catch (_) {
      return null;
    }
  }

  Future<List<FriendRemark>> getAllFriendRemarks(String userId) async {
    final resp = await http.get(
      Uri.parse('$_baseUrl/profile/$userId/friend-remarks'),
      headers: _headers,
    );
    return (jsonDecode(resp.body)['data'] as List)
        .map((e) => FriendRemark.fromJson(e))
        .toList();
  }

  // 好友分组
  Future<FriendGroup> createFriendGroup(String userId, String name, {int sortOrder = 0}) async {
    final resp = await http.post(
      Uri.parse('$_baseUrl/profile/$userId/friend-group'),
      headers: _headers,
      body: jsonEncode({'name': name, 'sortOrder': sortOrder}),
    );
    return FriendGroup.fromJson(jsonDecode(resp.body)['data']);
  }

  Future<List<FriendGroup>> getFriendGroups(String userId) async {
    final resp = await http.get(
      Uri.parse('$_baseUrl/profile/$userId/friend-groups'),
      headers: _headers,
    );
    return (jsonDecode(resp.body)['data'] as List)
        .map((e) => FriendGroup.fromJson(e))
        .toList();
  }

  Future<void> deleteFriendGroup(String userId, String groupId) async {
    await http.delete(
      Uri.parse('$_baseUrl/profile/$userId/friend-group/$groupId'),
      headers: _headers,
    );
  }
}
